### PR TITLE
refactor: Remove unused and fragile string interface from arith_uint256

### PR DESCRIPTION
**Pull request by maflcko**
The string interface (`base_uint(const std::string&)`, as well as `base_uint::SetHex`) is problematic for many reasons:

* It is unused (except in test-only code).
* It is redundant with the `uint256` string interface: `std::string -> uint256 -> UintToArith256`.
* It is brittle, because it inherits the brittle `uint256` string interface, which is brittle due to the use of `c_str()` (embedded null will be treated as end-of string), etc ...

Instead of fixing the interface, remove it since it is unused and redundant with `UintToArith256`.

### CODE REVIEW

File: src/test/arith_uint256_tests.cpp [hunk 1]
```diff
@@ -95,25 +96,25 @@ BOOST_AUTO_TEST_CASE( basics ) // constructors, equality, inequality
     BOOST_CHECK(ZeroL == (OneL << 256));
 
     // String Constructor and Copy Constructor
```
Comment (TheCharlatan) [inline src/test/arith_uint256_tests.cpp:98]: This comment should be adjusted, since there is no `arith_uint256(const std::string& str)` constructor anymore.

File: src/test/uint256_tests.cpp [hunk 2]
```diff
@@ -279,6 +279,34 @@ BOOST_AUTO_TEST_CASE( operator_with_self )
     BOOST_CHECK(v == UintToArith256(uint256S("0")));
 }
 
+BOOST_AUTO_TEST_CASE(parse)
```
Comment (TheCharlatan) [inline src/test/uint256_tests.cpp:282]: Nit: Put these extra tests in a separate commit?

File: src/test/arith_uint256_tests.cpp [hunk 3]
```diff
@@ -95,25 +96,25 @@ BOOST_AUTO_TEST_CASE( basics ) // constructors, equality, inequality
...
-    BOOST_CHECK(R1L == arith_uint256(R1ArrayHex));
+    BOOST_CHECK(arith_uint256S("0x" + R1L.ToString()) == R1L);
+    BOOST_CHECK(arith_uint256S("0x" + R2L.ToString()) == R2L);
+    BOOST_CHECK(arith_uint256S("0x" + ZeroL.ToString()) == ZeroL);
+    BOOST_CHECK(arith_uint256S("0x" + OneL.ToString()) == OneL);
+    BOOST_CHECK(arith_uint256S("0x" + MaxL.ToString()) == MaxL);
+    BOOST_CHECK(arith_uint256S(R1L.ToString()) == R1L);
+    BOOST_CHECK(arith_uint256S("   0x" + R1L.ToString() + "   ") == R1L);
+    BOOST_CHECK(arith_uint256S("") == ZeroL);
+    BOOST_CHECK(R1L == arith_uint256S(R1ArrayHex));
```
Comment (TheCharlatan) [inline src/test/arith_uint256_tests.cpp:107]: These same tests also exist in `uint256_tests`. What are these additionally testing now? Can they be removed?

File: src/test/uint256_tests.cpp (hunk 2 above)
Comment (TheCharlatan) [reply src/test/uint256_tests.cpp:282]: ^^ I ran HEAD~2 by mistake when reviewing this.

File: src/test/arith_uint256_tests.cpp (hunk 3 above)
Comment (TheCharlatan) [reply src/test/arith_uint256_tests.cpp:107]: They're not really testing `arith_uint256` anymore though, just the new conversion helper. I don't feel strongly about this, these are just a bunch of test lines after all, but was curious what the criteria was here.

**Review by sedited — APPROVED** [review]
ACK fa63f16018d9468e1751d2107b5102184ac2d5ae

**Comment by ajtowns** [comment]
ACK fa63f16018d9468e1751d2107b5102184ac2d5ae
